libsolidity/codegen/YulUtilFunctions.cpp

```diff
@@ -3274,6 +3282,67 @@ string YulUtilFunctions::conversionFunction(Type const& _from, Type const& _to)
 	});
 }
 
+string YulUtilFunctions::bytesToFixedBytesConversionFunction(ArrayType const& _from, FixedBytesType const& _to)
+{
+	solAssert(_from.isByteArray(), "");
+	solAssert(_from.isDynamicallySized(), "");
+	string functionName = "convert_bytes_to_bytesNN_from_" + _from.identifier() + "_to_" + _to.identifier();
+	return m_functionCollector.createFunction(functionName, [&](auto& _args, auto& _returnParams) {
+		_args = { "array" };
+		bool fromCalldata = _from.dataStoredIn(DataLocation::CallData);
+		if (fromCalldata)
+			_args.emplace_back("len");
+		_returnParams = {"value"};
+		Whiskers templ(R"(
+			let length := <arrayLen>(array<?fromCalldata>, len</fromCalldata>)
+			if gt(length, <fixedBytesLen>) { <panic>() }
+			let dataArea := array
+			<?fromMemory>
+				dataArea := <dataArea>(array)
+			</fromMemory>
+			<?fromStorage>
+				if eq(length, 32) { dataArea := <dataArea>(array) }
+			</fromStorage>
+
+			<?fromCalldata>
+				value := <cleanup>(calldataload(dataArea))
+			<!fromCalldata>
+				value := <extractValue>(dataArea)
+			</fromCalldata>
+
+			if lt(length, <fixedBytesLen>) {
```

Comment:
Can't we just `and` with a compile-time constant mask in all cases?

Reply:
Or in other words: just call the cleanup function?

Reply:
This last part is if length of `bytes` array that is being converted is smaller than number of bytes in `bytesNN` we convert to.
For example:
```
function fromSlice(bytes calldata c) external returns (bytes8) {
    return bytes8(c[0:6]);
}
```
Since `bytes` is dynamically sized array we need to do this in runtime...
Also, this is for the case when we have storage array of length 31 and we convert to `bytes32`.
Similar for memory, we can find dirty bytes... I added the test for this cases in `bytes_to_fixed_bytes_cleanup.sol`

Reply:
Ah right, I confused it!

Then: shouldn't we also clean if length is equal to fixedBytesLen, because of potential dirty data? I think this is needed if we are using calldata slices.

Reply:
I had that problem with calldata, and that is why I call `cleanup` in case of calldata.
`readFromMemoryFunction` does cleanup itself